sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java

```diff
@@ -483,41 +532,115 @@ public void validate(PipelineOptions options) {
+        jobIdTokenView = jobIdTokenCollection
+            .apply("ViewId", View.<String>asSingleton());
+      }
 
-      if (getQuery() != null
-          && (!getQuery().isAccessible() || !Strings.isNullOrEmpty(getQuery().get()))) {
-        source =
-            BigQueryQuerySource.create(
-                stepUuid,
-                getQuery(),
-                getFlattenResults(),
-                getUseLegacySql(),
-                getBigQueryServices());
+      PCollection<TableRow> rows;
+      if (!getUseNewSource()) {
+        // Apply the traditional Source model.
+        rows = p.apply(
+            org.apache.beam.sdk.io.Read.from(applySourceTransform(staticJobUuid)))
+            .setCoder(getDefaultOutputCoder());
       } else {
-        source =
-            BigQueryTableSource.create(
-                stepUuid,
-                getTableProvider(),
-                getBigQueryServices());
+        // Apply the DoFn version.
+        final TupleTag<String> filesTag =
+            new TupleTag<String>(){};
+        final TupleTag<String> tableSchemaTag =
+            new TupleTag<String>(){};
+        PCollectionTuple tuple = jobIdTokenCollection.apply("RunCreateJob", ParDo
+            .of(
+                new DoFn<String, String>() {
+                  @ProcessElement
+                  public void processElement(ProcessContext c)
+                      throws Exception {
+                    String jobUuid = c.element();
+                    BigQuerySourceBase source = applySourceTransform(
+                        jobUuid);
+                    String schema = BigQueryHelpers.toJsonString(
+                        source.getSchema(c.getPipelineOptions()));
+                    c.output(tableSchemaTag, schema);
+                    List<ResourceId> files = source.extractFiles(c.getPipelineOptions());
+                    for (ResourceId file : files) {
+                      c.output(file.toString());
+                    }
+                  }
+                })
+            .withOutputTags(filesTag, TupleTagList.of(tableSchemaTag)));
+        final PCollectionView<String> schemaView = tuple.get(tableSchemaTag)
+            .apply(View.<String>asSingleton());
+        rows = tuple.get(filesTag)
+            .apply(WithKeys.of(new SerializableFunction<String, String>() {
+                public String apply(String s) {
+                  return s;
+                }
+              }))
+            .apply(Reshuffle.<String, String>of())
+            .apply(Values.<String>create())
+            .apply("ReadFiles", ParDo.of(
+                new DoFn<String, TableRow>() {
+                  @ProcessElement
+                  public void processElement(ProcessContext c)
+                      throws Exception {
+                    TableSchema schema = BigQueryHelpers.fromJsonString(
+                        c.sideInput(schemaView), TableSchema.class);
+                    String jobUuid = (String) c.sideInput(jobIdTokenView);
```

Comment:
Still grumps at me after moving:

ThroughThenCleanup.java:[103,39] incompatible types: java.lang.Object cannot be converted to java.lang.String